experimental/overhead_matching/kimera_spectacular_data_provider.cc

```diff
@@ -0,0 +1,282 @@
+        LOG(WARNING) << "Provided final frame for KimeraSpectacularDataProvider was above the "
+                        "number of frames. Reducing to match size of dataset";
+        final_k_ = spec_log_.num_frames();
+    }
+
+    CHECK_GT(final_k_, 0);
+
+    CHECK_GT(final_k_, initial_k_)
+        << "Value for final_k (" << final_k_ << ") is smaller than value for"
+        << " initial_k (" << initial_k_ << ").";
+    current_k_ = initial_k_;
+}
+
+/* -------------------------------------------------------------------------- */
+SpectacularDataProviderInterface::SpectacularDataProviderInterface(const VIO::VioParams& vio_params)
+    : SpectacularDataProviderInterface(FLAGS_dataset_path, FLAGS_initial_k, FLAGS_final_k,
+                                       vio_params) {}
+
+/* -------------------------------------------------------------------------- */
+SpectacularDataProviderInterface::~SpectacularDataProviderInterface() {
+    LOG(INFO) << "SpectacularDataProviderInterface destructor called.";
+}
+
+/* -------------------------------------------------------------------------- */
+bool SpectacularDataProviderInterface::spin() {
+    if (!is_imu_data_sent_) {
+        // First, send all the IMU data. The flag is to avoid sending it several
+        // times if we are running in sequential mode.
+        if (imu_single_callback_) {
+            sendImuData();
+        } else {
+            LOG(ERROR) << "Imu callback not registered! Not sending IMU data.";
+        }
+        is_imu_data_sent_ = true;
+    }
+
+    // Spin.
+    CHECK_GT(final_k_, initial_k_);
+    // We log only the first one, because we may be running in sequential mode.
+    LOG_FIRST_N(INFO, 1) << "Running dataset between frame " << initial_k_ << " and frame "
+                         << final_k_;
+    while (!shutdown_ && spinOnce()) {
+        if (!vio_params_.parallel_run_) {
+            // Return, instead of blocking, when running in sequential mode.
+            return true;
+        }
+    }
+    LOG_IF(INFO, shutdown_) << "shutdown requested.";
+    return false;
+}
+
+bool SpectacularDataProviderInterface::hasData() const { return current_k_ < final_k_; }
+
+/* -------------------------------------------------------------------------- */
+bool SpectacularDataProviderInterface::spinOnce() {
+    CHECK_LT(current_k_, std::numeric_limits<VIO::FrameId>::max())
+        << "Are you sure you've initialized current_k_?";
+    if (current_k_ >= final_k_) {
+        LOG(INFO) << "Finished spinning dataset.";
+        return false;
+    }
+    // TODO: How do we want to handle camera parameters?
+    //  const VIO::CameraParams& left_cam_info = vio_params_.camera_params_.at(0);
+    const VIO::CameraParams left_cam_FAKE_PARAMS;
+    const bool& equalize_image = false;
+    // vio_params_.frontend_params_.stereo_matching_params_.equalize_image_;
+
+    std::optional<robot::experimental::overhead_matching::FrameGroup> maybe_frame =
```

Comment:
Since this function is in the `robot::experimental::overhead_matching` namespace, this can just be `std::optional<FrameGroup>`.